tests/data/propchain_dep/params/hmodel/expected_predict_server.yaml

```diff
@@ -0,0 +1,70 @@
+responses:
+- _type: CheckAlignmentResponse
+  contents:
+    unknown_definitions:
+    - <pytact/graph_api.capnp:Definition reader (name = "Tactician.Ltac1.Record.private_constant_placeholder",
```

Comment:
If you are going to use yaml, you might as well print this as yaml too. You can get a dictionary version of the data using `d.lowlevel.to_dict()`, which you can then convert to json. Alternatively, you can also just print the node representation (like you do for arguments) using `repr(d.node)`. Or you can print the just the name of the definition using `d.node`.

Reply:
To be clear, in some sense this is just yaml.  This is just a string.  But yes, I would have like to have it broken up more.  Unfortionately, `d.lowlevel.to_dict()` and other approaches I tried, don't do it automatically.  I get errors about classes without `to_dict`.  So I just stopped at this point and took the representation.  The `repr(d.node())` approach is interesting.  I do have to admit I really like having the name here.  It makes it much easier to see what is going on especially if this test ever breaks at this point.

Reply:
@LasseBlaauwbroek I don't suppose there is a similar innovation to get the name of the tactic or the name of the local argument, is there?  That would make the output a lot more readable.

Reply:
Regarding local arguments: I just [pushed](https://github.com/coq-tactician/coq-tactician-reinforce/commit/413caa73aa607897f4ac9dbcabf911c3fa1cc161) a change with a new option `Set Tactician Neural Metadata`, which will force Coq to include all of the metadata that is also present in the dataset when transmitting data to the server. This can be pretty expensive, so you should only use it in test/debug situations. The name of a local argument `a : Node` can be found by finding `i` such that `proof_state.context[i] == a` and then finding its name using `proof_state.context_names[i]` and optionally it's type using `proof_state.context_text[i]`. Here `proof_state` refers to to object being cast on line 435:
```
if isinstance(msg, ProofState)
```

Regarding tactic names: Currently, the capnproto format doesn't allow for that. But we could certainly add this as an option in the future (please open an issue if you want that).

Reply:
(And yes, I know that this method of looking up names of local context elements is cumbersome. But this cannot be helped due to the graph sharing algorithm. Structurally equal hypotheses that are part of different proof states will be shared using the algorithm. But they can still have different names. So the node corresponding to a hypothesis doesn't have a unique name. Therefore, the name has to be stored in the proof_state object, where it is unique.